Add optional variable to Makefile.generated_files for auto-updating the API rule violation exception list

roycaihw opened the pull request:
The API rule violations reported by openapi-gen are in format of

 - `{rule-name},{package},{type},{(optional) field}`

Previous the report file was partially-sorted on {package} level and {type} level, which makes it hard for API authors to update the exception list when necessary, e.g. https://github.com/kubernetes/kubernetes/pull/60195#discussion_r204056998, https://github.com/kubernetes/kubernetes/pull/66722

This PR bumps the `k8s.io/kube-openapi` dependency. Now the generated report of API rule violations are sorted alphabetically on each of the {rule-name}, {package}, {type}, and {field} levels. 

This PR also adds an option to run 
```
make generated_files UPDATE_API_KNOWN_VIOLATIONS=true
```
which lets the generator to update the checked-in violation exceptions list automatically.

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

/sig api-machinery

roycaihw commented:
/kind cleanup

roycaihw commented on build/root/Makefile.generated_files:
```diff
@@ -405,7 +405,7 @@ OPENAPI_OUTPUT_PKG := pkg/generated/openapi
 BOILERPLATE_FILENAME := vendor/k8s.io/code-generator/hack/boilerplate.go.txt
 REPORT_FILENAME := $(OUT_DIR)/violations.report
 KNOWN_VIOLATION_FILENAME := api/api-rules/violation_exceptions.list
```
Currently the workflow is: API authors see the error message below when `violations.report` differs from `violation_exceptions.list`, and they must follow the instruction in the message to either 
- fix new API rule violations that they introduced
- run `hack/update-api-violation-exceptions.sh` to update the exception list when necessary

Alternatively we could have `REPORT_FILENAME := api/api-rules/violation_exceptions.list`, by which we can get rid of the hack script-- the API authors won't see any error when there are new violations, the `violation_exceptions.list` will be updated, and it becomes the API reviewers' responsibility to catch the file change and ask API authors to fix the violations.

roycaihw replied:
/cc @lavalamp 
for high level workflow review
/cc @ixdy 
for Makefile and hack script review

lavalamp replied:
I'm fine with either option--whatever @ixdy thinks is best.

roycaihw replied:
> Just generating the exceptions list automatically seems like the best experience for contributors. Do we expect api-approvers to notice changes to this list?

I suspect that's unlikely to happen. When api-authors add/update APIs and send a PR, I'd expect they put the auto-updated exceptions into a ["generated"](https://github.com/kubernetes/kubernetes/search?q=generated&type=Commits) commit, along with the other auto-generated files (bazel BUILD, openapi/swagger specs, api-reference docs, deepcopy, etc.). I think those commits usually are not (and shouldn't be) well-reviewed.

It would be an annoying extra step for api-reviewers to search for the exceptions filename in the "generated" commits, especially when most of the time that file shouldn't be changed.

> Yet another option would be to optionally make Makefile.generated_files update the known violations file

I love this idea! Thanks @ixdy. Let me try implementing it

roycaihw replied:
@ixdy Updated. Please take a look

jennybuckley commented:
/assign @lavalamp 
/assign @ixdy

ixdy commented on hack/update-api-violation-exceptions.sh:
```diff
@@ -0,0 +1,79 @@
+# because somehow it will invoke `make generated_files`, which verifies the
+# API rule violation exception list, and defeats the purpose of this hack
+# script.
+hack/make-rules/build.sh ./vendor/k8s.io/kube-openapi/cmd/openapi-gen
```
nit: prefix `hack/make-rules/build.sh` with `${KUBE_ROOT}`; i.e. make this line
```shell
"${KUBE_ROOT}/hack/make-rules/build.sh" ./vendor/k8s.io/kube-openapi/cmd/openapi-gen
```

ixdy commented:
/lgtm
/approve

I assume you've verified this does the right thing in both modes?

ixdy commented on api/api-rules/README.md:
```diff
@@ -12,11 +12,20 @@ e.g.
+If you're removing violations from the exception list, or if there are good
+reasons for new APIs to violate API rules, please update the file using:
+
+ - `make UPDATE_API_KNOWN_VIOLATIONS=true generated_files`
```
I think usually we've put the variable after the target, e.g. `make generated_files UPDATE_API_KNOWN_VIOLATIONS=true` but I'm not sure if it makes a difference in practice.

roycaihw commented:
@ixdy Updated to put the variable after the target.

> I assume you've verified this does the right thing in both modes?

Yes. Without the last commit (which updates the list):
```
$ make clean
$ make generated_files
... files differ; error message printed
$ make generated_files
... make target cached, no error message shown, the exceptions list unchanged
$ make generated_files UPDATE_API_KNOWN_VIOLATIONS=true
... the exceptions list updated
```

cc @lavalamp for approval

roycaihw commented:
/retest

lavalamp commented:
I actually realized last week while doing something mostly unrelated that the linter doesn't use gengo's built-in assemble/verify functionality. I will send a demonstration PR to kube-openapi for Haowei to look at. This may not be necessary if Haowei likes the way that looks.

lavalamp commented:
https://github.com/kubernetes/kube-openapi/pull/111

lavalamp commented:
So, 111 has merged. Can we simplify this now and just rely on the --verify-only flag to keep it in sync?

roycaihw commented:
@lavalamp Hit one problem with the `--verify-only` approach: 

The flag also triggers the openAPIGen to verify file (`zz_generated.openapi.go` in openAPIGen's case). However this file is not checked-in, and also gets removed by `make clean`, so the following error occurs
```
unable to read file "openapi/zz_generated.openapi.go" for comparison
```

I suggest we can disable the default [VerifyFile](https://github.com/kubernetes/gengo/blob/5b57d243f2ca39dbbda758ee07b76b4d519f6dc7/generator/execute.go#L93) for openAPIGen by wrapping a CustomFileType that no-ops on VerifyFile (with documentation), since currently no one is using this functionality.

lavalamp commented:
Yeah let's not start using --verify-only mode just yet :/

roycaihw committed:
Add UPDATE_API_KNOWN_VIOLATIONS=true to Makefile.generated_files

One can run:
    make generated_files UPDATE_API_KNOWN_VIOLATIONS=true
to update the checked-in known API violation exceptions list.

Also updated README and the error message on how to resolve the
failure.

roycaihw commented:
@lavalamp Updated the error message and README. Please take a look.

A example of the error message:

```
2018/11/01 17:35:16 Code for OpenAPI definitions generated
1d0
< API rule violation: names_match,k8s.io/api/authorization/v1beta1,SubjectAccessReviewSpec,Groups
ERROR: 
	 API rule check failed. Reported violations differ from known violations. Please read api/api-rules/README.md to resolve the failure. 

Makefile.generated_files:453: recipe for target 'pkg/generated/openapi/zz_generated.openapi.go' failed
```

roycaihw commented:
/retest

lavalamp commented:
/lgtm
/approve

lavalamp commented:
@BenTheElder @ixdy @liggitt the ci robot doesn't check permissions correctly. I'm technically not in the api-approvers team at the moment, yet it applied the label when I approved.

ixdy commented:
It looks like you are:

https://github.com/kubernetes/kubernetes/blob/69f5f5eff2803317c794bb545705185fed74c201/OWNERS_ALIASES#L193-L198

lavalamp commented:
Yeah sorry for the false alarm, I thought the aliases referred to the github teams of the same name.
